localstack/services/s3/provider_stream.py

```diff
@@ -0,0 +1,754 @@
+                prefix_including_delimiter = name[0:end_of_delimiter]
+                common_prefixes.append(prefix_including_delimiter)
+                continue
+
+            # Differentiate between FakeKey and FakeDeleteMarkers
+            if isinstance(version, s3_models.FakeDeleteMarker):
+                delete_markers.append(version)
+                continue
+
+            requested_versions.append(version)
+
+        common_prefixes = sorted(set(common_prefixes))
+
+        return requested_versions, common_prefixes, delete_markers
+
+    @patch(s3_responses.S3Response.key_response)
+    def _fix_key_response(fn, self, *args, **kwargs):
+        """Return an iterator from the stream value in `key_value`"""
+        status_code, resp_headers, key_value = fn(self, *args, **kwargs)
+        content = get_generator_from_stream(key_value)
+        return status_code, resp_headers, content
+
+    @patch(s3_responses.S3Response._handle_range_header, pass_target=False)
+    def _handle_range_header(
+        self, request: Any, response_headers: dict[str, Any], response_content: Any
+    ) -> TYPE_RESPONSE:
+        is_streamed = isinstance(response_content, SpooledTemporaryFile)
+        if is_streamed:
+            response_content.seek(0, 2)
+            length = response_content.tell()
+            response_content.seek(0)
+        else:
+            length = len(response_content)
+
+        last = length - 1
+        _, rspec = request.headers.get("range").split("=")
+        if "," in rspec:
+            raise NotImplementedError("Multiple range specifiers not supported")
+
+        def toint(i: Any) -> Optional[int]:
+            return int(i) if i else None
+
+        begin, end = map(toint, rspec.split("-"))
+        if begin is not None:  # byte range
+            end = last if end is None else min(end, last)
+        elif end is not None:  # suffix byte range
+            begin = length - min(end, length)
+            end = last
+        else:
+            return 400, response_headers, ""
+        if begin < 0 or end > last or begin > min(end, last):
+            raise s3_exceptions.InvalidRange(
+                actual_size=str(length), range_requested=request.headers.get("range")
+            )
+        response_headers["content-range"] = f"bytes {begin}-{end}/{length}"
+
+        if not is_streamed:
+            content = response_content[begin : end + 1]
+            response_headers["content-length"] = len(content)
+        else:
+            stream = PartialStream(response_content, begin, end)
+            content = get_generator_from_stream(stream)
+            requested_length = end - begin + 1
+            response_headers["content-length"] = requested_length
+
+        return 206, response_headers, content
+
+
+def get_generator_from_stream(response_content: Any):
+    # Werkzeug will only read 1 everytime, so we control how much we return
+    if isinstance(response_content, (SpooledTemporaryFile, PartialStream)):
```

Comment:
todo: create a special generator for stream request instead of using PartialStream, and keep the position for regular streams